16787: allow Dir.home to work for non-login procs when $HOME not set

salewski opened the pull request:
Allow the 'Dir.home' method has to reliably locate the user's home directory
when all three of the following are true at the same time:

    1. Ruby is running on a Unix-like OS
    2. The $HOME environment variable is not set
    3. The process is not a descendant of login(1) (or a work-alike)

This is accomplished by looking up the user's record in the password database
by uid (getpwuid_r(3)) rather than by the name returned by getlogin(3).

salewski commented:
The entry for this issue in the bug database is:

    https://bugs.ruby-lang.org/issues/16787

I see that there are some issues with the automated tests. I'll see what I can do about getting those cleaned up.

nobu commented:
Any particular reason to remove the current fallback if `getpwuid_r` is unavailable?

salewski commented:
> Any particular reason to remove the current fallback if `getpwuid_r` is unavailable?

Hi @nobu; thanks for the feedback. If we want to have a fallback for `getpwuid_r`, I think it should be `getpwuid` rather than `getlogin` for two reasons:

   1. Functionality-wise it is similar to `getpwuid_r` in that it, too, works for non-login processes.

   2. Portability-wise I would expect it to be more readily available than `getlogin`. According to the `HISTORY` sections of the NetBSD manpages for [getpwuid](https://netbsd.gw.com/cgi-bin/man-cgi?getpwuid++NetBSD-9.0) and [getlogin](https://netbsd.gw.com/cgi-bin/man-cgi?getlogin++NetBSD-9.0), the former appeared in Version 7 AT&T UNIX ([circa 1979](https://en.wikipedia.org/wiki/Version_7_Unix)) and conforms to ISO/IEC 9945-1:1990 ("POSIX.1"), but the latter did not appear until 4.4BSD ([circa 1994](https://en.wikipedia.org/wiki/History_of_the_Berkeley_Software_Distribution)).

As for the `getuid` function, I'm not worried about that one because it [first appeared in Version 1 AT&T UNIX](https://netbsd.gw.com/cgi-bin/man-cgi?getuid++NetBSD-current) (which [is ancient](https://en.wikipedia.org/wiki/Ancient_UNIX), [circa 1971](https://en.wikipedia.org/wiki/Research_Unix)  :-)

I should also add that I *do* think using a fallback to `getpwuid_r` is warranted, as you suggest. In NetBSD the function [first appeared in NetBSD 3.0](https://netbsd.gw.com/cgi-bin/man-cgi?getpwuid_r++NetBSD-9.0), which [was released in 2005](https://en.wikipedia.org/wiki/NetBSD). From a portability perspective, I think that is **"too recent"** to depend on. What do you think?

Also, if we are going to use a fallback for `getpwuid_r`, then it might make sense to also loosen the dependency that I introduced on the `sysconf` definition of `_SC_GETPW_R_SIZE_MAX`. While it dates back to IEEE Std 1003.1c-1995 ("POSIX.1"), I'm not certain that its presence on historical machines is always paired with the availability of the `getpwuid_r` function.

Let me rework the patch to see what it looks like. Thanks again, and more soon...

salewski commented:
> > Any particular reason to remove the current fallback if `getpwuid_r` is unavailable?
> 
> [...]If we want to have a fallback for `getpwuid_r`, I think it should be `getpwuid` rather than `getlogin` for two reasons:
> [...]

After pushing my changes that do the `getpwuid_r` fallback to `getpwud` thing, I saw [your comment](https://bugs.ruby-lang.org/issues/16787?issue_count=2&issue_position=1&next_issue_id=12226#note-2) in the bug tracker about the possibility of multiple login names being mapped to the same uid, possibly with different default home directories.

Thank you for that insight, @nobu. I had not considered that angle, but now that you have pointed it out to me I can see that falling back from name-based lookups (`getpwname_r`, `getpwname`) to uid-based lookups (`getpwuid_r`, `getpsuid`) is the only way to go if we want to avoid breaking those use cases.

Let me rework the patch some more and then I'll circle back.

salewski commented:
> > > Any particular reason to remove the current fallback if `getpwuid_r` is unavailable?
> > 
> > [...]
> 
> [note about the possibility of multiple login names being mapped to the same uid, possibly with different default home directories.]
> 
> Thank you for that insight, @nobu.
> [...]
> 
> Let me rework the patch some more and then I'll circle back.

Hi @nobu, How do the changes in [commit 3a54183d3ae1][1] look to you?

The changes preserve the default behavior of a doing name-based lookup first, and then falls back on doing the lookup by uid only if the name based lookup fails, as you suggested.

The changes in the above commit are on my [ads/b.r-l.o-issue-16787][2] branch, which was rebased on top of the HEAD of `master` earlier today. There is [a single failing integration test][3] (1 out of 95), as reported by the automated checks. It looks to me that the [two][4] [jobs][5] are perhaps not happy with the rebasing/forced push to my branch. Is there anything I can do/need to do to make appveyor happy?


[1]: https://github.com/ruby/ruby/commit/3a54183d3ae15896b3c784b46d5b859156b7f5e5    "commit: 3a54183d3ae1"
[2]: https://github.com/salewski/ruby/tree/ads/b.r-l.o-issue-16787                   "branch: ads/b.r-l.o-issue-16787"
[3]: https://ci.appveyor.com/project/ruby/ruby/builds/32231996                       "url: appveyor failure for build 32231996"
[4]: https://ci.appveyor.com/project/ruby/ruby/builds/32231996/job/wu7cxkl7d3hf9cvd  "appveyor build 32231996, job wu7cxkl7d3hf9cvd"
[5]: https://ci.appveyor.com/project/ruby/ruby/builds/32231996/job/yg3uesa9rphg776q  "appveyor build 32231996, job yg3uesa9rphg776q"

nobu commented on file.c:
```diff
@@ -3596,34 +3631,298 @@ rb_home_dir_of(VALUE user, VALUE result)
+    errno = 0;
+    while ((gle = getlogin_r(login, loginsize)) != 0) {
+
+        if (gle == ENOTTY || gle == ENXIO || ENOENT) {
```
Mere `ENOENT` without comparison seems wrong.

salewski replied:
Indeed; thanks for the catch.

nobu commented on file.c:
```diff
@@ -3596,34 +3631,298 @@ rb_home_dir_of(VALUE user, VALUE result)
+#  error BUG: Either USE_GETLOGIN_R or USE_GETLOGIN should be defined here.
+#  error If you encounter this message, can you file a bug report?
+#  error Remember to attach a detailed description of your environment.
+#  error Thank you!
```
If this is ordered as `#if USE_GETLOGIN_R`, `#elif USE_GETLOGIN` and `#else`, these `#error`s would be unnecessary.

salewski replied:
Yes, they are ordered that way, and those conditional bits are also more localized than when I first introduced the `#error` blocks -- now they're excessive. Thanks; I'll remove them.

nobu commented on file.c:
```diff
@@ -3596,34 +3631,298 @@ rb_home_dir_of(VALUE user, VALUE result)
+        return Qnil;
+    }
+
+    VALUE result = rb_str_new_cstr(login);
```
No need to create a new `String` from `login`, it is a pointer into a `String`.

salewski replied:
Roger that; thanks.

nobu commented on file.c:
```diff
@@ -3596,34 +3631,298 @@ rb_home_dir_of(VALUE user, VALUE result)
+        return result;
+    }
+
+    rb_raise(rb_eArgError, "couldn't find home for uid `%ld'", (long)getpid());
```
Preferably, the error messages should be kept for each case, for the backward compatibilities.
And some compilers don't recognize `NORETURN` function but warn that no return at the end.
`return` for `getenv("HOME")` is placed there for that reason.

salewski replied:
> Preferably, the error messages should be kept for each case, for the backward compatibilities.

With the code always falling back on doing the lookup by uid, the case for the first error message will never happen:
```ruby
           rb_raise(rb_eArgError, "couldn't find HOME for login `%s' -- expanding `~'",
                    login);
```

But I'll put the bottoming-out error message back to:
```
           rb_raise(rb_eArgError, "couldn't find login name -- expanding `~'");
```
I had changed that because the specific "`...expanding `~'...`" wording was a bit of a misnomer, and with the lookup by uid even more so. But I see your point about backward compatibility in this regard.


> And some compilers don't recognize `NORETURN` function but warn that no return at the end.
> `return` for `getenv("HOME")` is placed there for that reason.

Ah, I see. I'll have to account for that type of thing in a couple of places with my changes.

Thanks for all the great feedback, @nobu. I'll incorporate these changes, and will also look to see about whether pieces of them might live more comfortably on in `process.c`, [as you mentioned in the issue tracker](https://bugs.ruby-lang.org/issues/16787?issue_count=2&issue_position=1&next_issue_id=12226#note-11).

salewski commented:
Hi @nobu, I left detailed comments about this set of changes [in the comments](https://bugs.ruby-lang.org/issues/16787#note-13) in the bug tracker, but the short version is that I addressed all of the items from your earlier review (thanks for that!), moved the three new functions to `process.c` as you suggested, did  a small amount of additional tidying up, and did some more testing. Please take a look when you can and let me know if there are any further changes you would like to see, as well. Thanks!

salewski commented:
Hi @nobu, I just refreshed the pull request (rebased on top of current `master` (2020-05-14)), but otherwise did not make any changes beyond those noted earlier. Please take a look when you can and let me know if there are any further changes you would like to see, as well. Thanks!

salewski commented:
Hi @nobu, Are there any further changes you would like to see with this PR?

Thanks,
-Al
